File: Week6 (Graph [SSSP, APSP, BFS, DFS])/All Pair Shortest Path (APSP)/apsp.c
#include <stdio.h>
#include <limits.h>

#define INF 1000000000

void extendSP(int L_r_1[][4], int W[][4], int L_r[][4], int n) {
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            for (int k = 0; k < n; k++) {
                if (L_r[i][j] > L_r_1[i][k] + W[k][j]) {
                    L_r[i][j] = L_r_1[i][k] + W[k][j];
                }
            }
        }
    }
}

void slowAPSP(int W[][4], int L_0[][4], int n, int L[][4]) {
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            L[i][j] = L_0[i][j];
        }
    }

    for (int r = 0; r < n - 1; r++) {
        int M[4][4];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                M[i][j] = INF;
            }
        }
        extendSP(L, W, M, n);
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                L[i][j] = M[i][j];
            }
        }
    }
}

void fasterAPSP(int W[][4], int n, int L[][4]) {
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            L[i][j] = W[i][j];
        }
    }

    int r = 1;
    while (r < n - 1) {
        int M[4][4];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                M[i][j] = INF;
            }
        }
        extendSP(L, L, M, n);
        r *= 2;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                L[i][j] = M[i][j];
            }
        }
    }
}

int main() {
    int n = 4;
    int W[4][4] = {
        {0, 3, 8, INF},
        {INF, 0, 4, 11},
        {INF, INF, 0, 7},
        {4, INF, INF, 0}
    };
    int L_0[4][4] = {
        {0, INF, INF, INF},
        {INF, 0, INF, INF},
        {INF, INF, 0, INF},
        {INF, INF, INF, 0}
    };

    int result_slow[4][4];
    slowAPSP(W, L_0, n, result_slow);
    printf("[Slow] All Pairs Shortest Path matrix:\n");
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            if (result_slow[i][j] == INF) {
                printf("INF ");
            } else {
                printf("%d ", result_slow[i][j]);
            }
        }
        printf("\n");
    }

    int result_faster[4][4];
    fasterAPSP(W, n, result_faster);
    printf("\n[Faster] All Pairs Shortest Path matrix:\n");
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            if (result_faster[i][j] == INF) {
                printf("INF ");
            } else {
                printf("%d ", result_faster[i][j]);
            }
        }
        printf("\n");
    }

    return 0;
}

File: Week6 (Graph [SSSP, APSP, BFS, DFS])/All Pair Shortest Path (APSP)/apsp.py
INF = 1e+9

def extendSP(L_r_1, W, L_r, n: int):
    for i in range(n):
        for j in range(n):
            for k in range(n):
                if L_r[i][j] > L_r_1[i][k] + W[k][j]:
                    L_r[i][j] = L_r_1[i][k] + W[k][j]

def slowAPSP(W, L_0, n: int):
    L = [[INF for i in range(n)] for i in range(n)]

    L = L_0
    for r in range(n-1):
        M = [[INF for i in range(n)] for i in range(n)]
        extendSP(L, W, M, n)
        L = M
    
    return L

def fasterAPSP(W, n: int):
    L = [[INF for i in range(n)] for i in range(n)]

    L = W
    r = 1
    while r < n-1:
        M = [[INF for i in range(n)] for i in range(n)]
        extendSP(L, L, M, n)
        r *= 2
        L = M
    
    return L

def main():
    n = 4
    W = [
        [0, 3, 8, INF],
        [INF, 0, 4, 11],
        [INF, INF, 0, 7],
        [4, INF, INF, 0]
    ]
    L_0 = [
        [0, INF, INF, INF],
        [INF, 0, INF, INF],
        [INF, INF, 0, INF],
        [INF, INF, INF, 0]
    ]

    result = slowAPSP(W, L_0, n)
    print("[Slow] All Pairs SHortest Path matrix:")
    for i in range(n):
        for j in range(n):
            if result[i][j] == INF:
                print("INF", end=' ')
            else:
                print(result[i][j], end=' ')
        print()
    
    print()
    result = fasterAPSP(W, n)
    print("[Faster] All Pairs SHortest Path matrix:")
    for i in range(n):
        for j in range(n):
            if result[i][j] == INF:
                print("INF", end=' ')
            else:
                print(result[i][j], end=' ')
        print()


if __name__ == "__main__":
    main()